swaps/swaps_service.go

```diff
@@ -125,254 +183,1057 @@ func (svc *swapsService) EnableAutoSwaps(ctx context.Context, lnClient lnclient.
+		_ = boltzWs.Close()
+		svc.markSwapState(&dbSwap, constants.SWAP_STATE_FAILED)
+		return nil, err
+	}
+
+	go svc.startSwapInListener(&dbSwap, boltzWs)
+
+	return &SwapResponse{
+		SwapId:      swap.Id,
+		PaymentHash: invoice.PaymentHash,
+	}, nil
+}
+
+func (svc *swapsService) CalculateSwapOutFee() (*SwapFees, error) {
+	reversePairs, err := svc.boltzApi.GetReversePairs()
+	if err != nil {
+		return nil, fmt.Errorf("could not get reverse pairs: %s", err)
+	}
+
+	pair := boltz.Pair{From: boltz.CurrencyBtc, To: boltz.CurrencyBtc}
+	pairInfo, err := boltz.FindPair(pair, reversePairs)
+	if err != nil {
+		return nil, fmt.Errorf("could not find reverse pair: %s", err)
+	}
+
+	fees := pairInfo.Fees
+	networkFee := fees.MinerFees.Lockup + fees.MinerFees.Claim
+
+	return &SwapFees{
+		AlbyServiceFee:  AlbySwapServiceFee,
+		BoltzServiceFee: fees.Percentage,
+		BoltzNetworkFee: networkFee,
+	}, nil
+}
+
+func (svc *swapsService) CalculateSwapInFee() (*SwapFees, error) {
+	submarinePairs, err := svc.boltzApi.GetSubmarinePairs()
+	if err != nil {
+		return nil, fmt.Errorf("could not get reverse pairs: %s", err)
+	}
+
+	pair := boltz.Pair{From: boltz.CurrencyBtc, To: boltz.CurrencyBtc}
+	pairInfo, err := boltz.FindPair(pair, submarinePairs)
+	if err != nil {
+		return nil, fmt.Errorf("could not find reverse pair: %s", err)
+	}
+
+	fees := pairInfo.Fees
+
+	return &SwapFees{
+		AlbyServiceFee:  AlbySwapServiceFee,
+		BoltzServiceFee: fees.Percentage,
+		BoltzNetworkFee: fees.MinerFees,
+	}, nil
+}
+
+func (svc *swapsService) markSwapState(dbSwap *db.Swap, state string) {
+	if svc.db.Limit(1).Find(dbSwap, &db.Swap{
+		SwapId: dbSwap.SwapId,
+		State:  state,
+	}).RowsAffected > 0 {
+		logger.Logger.WithField("swapId", dbSwap.SwapId).Debugf("swap already marked as %s", state)
+		return
+	}
+
+	dbErr := svc.db.Model(dbSwap).Updates(&db.Swap{
+		State: state,
+	}).Error
+	if dbErr != nil {
+		logger.Logger.WithError(dbErr).WithField("swapId", dbSwap.SwapId).Error("Failed to update swap state")
+	}
+}
+
+func (svc *swapsService) RefundSwap(swapId, address string) error {
+	var swap db.Swap
+	err := svc.db.Limit(1).Find(&swap, &db.Swap{
+		SwapId: swapId,
+	}).Error
 	if err != nil {
-		return fmt.Errorf("could not create swap: %s", err)
+		logger.Logger.WithField("swapId", swapId).WithError(err).Error("Could not find swap to process refund")
+		return err
+	}
+
+	if swap.Type != constants.SWAP_TYPE_IN {
+		return fmt.Errorf("cannot process refund for swap id: %s", swapId)
 	}
 
-	boltzPubKey, err := btcec.ParsePubKey(swap.RefundPublicKey)
+	if swap.ClaimTxId != "" {
+		return fmt.Errorf("refund already processed with claim txid: %s", swap.ClaimTxId)
+	}
+
+	network, err := boltz.ParseChain(svc.cfg.GetNetwork())
 	if err != nil {
 		return err
 	}
 
-	tree := swap.SwapTree.Deserialize()
-	if err := tree.Init(boltz.CurrencyBtc, true, ourKeys, boltzPubKey); err != nil {
+	swapTransactionResp, err := svc.boltzApi.GetSwapTransaction(swapId)
```

Comment:
is it good to rely on Boltz here? shouldn't we know this already from our own swaps DB? (maybe we can add a TODO to revisit this?)

Reply:
We are mainly doing this to get the lockup transaction hex which is used to construct the tx below (and we can't assume to have it always since it can be done from external wallet too), we can fetch this (if `swap.LockupTxId != ""` using mempool api too somehow but I don't see much difference. But I'll add this as a comment for reference as it looks as if we are doing it for txId

Reply:
Ah ok, thanks!